src/middlewared/middlewared/plugins/directoryservices.py

```diff
@@ -110,6 +176,126 @@ async def nss_info_choices(self, dstype):
 
         return ret
 
+    @private
+    def backup_secrets(self):
+        ha_mode = self.middleware.call_sync('smb.get_smb_ha_mode')
+        netbios_name = self.middleware.call_sync('smb.config')['netbiosname_local']
+        db = self.middleware.call_sync('datastore.query',
+                                       'services.cifs', [],
+                                       {'prefix': 'cifs_srv_', 'get': True})
+
+        if db['secrets'] is None:
+            db_secrets = {}
+        else:
+            db_secrets = json.loads(db['secrets'])
+
+        with DirectorySecrets(logger=self.logger, ha_mode=ha_mode) as s:
+            secrets = s.dump()
+
+        if not secrets:
+            self.logger.warning("Unable to parse secrets")
+            return
+
+        db_secrets.update({netbios_name: secrets})
+        self.middleware.call_sync('datastore.update',
+                                  'services.cifs', 1,
```

Comment:
We should be using the id from `db` as we might have multiple rows in config services in the db and when we fix that, it will be easier to track changes and ensure there are no regressions.

Reply:
Good point